cli/npm/managed/resolvers/local/bin_entries.rs

```diff
@@ -3,7 +3,201 @@
+  ) -> Result<(), AnyError> {
+    if !self.entries.is_empty() && !bin_node_modules_dir_path.exists() {
+      std::fs::create_dir_all(bin_node_modules_dir_path).with_context(
+        || format!("Creating '{}'", bin_node_modules_dir_path.display()),
+      )?;
+    }
+
+    if self.has_collision {
+      // walking the dependency tree to find out the depth of each package
+      // is sort of expensive, so we only do it if there's a collision
+      sort_by_depth(snapshot, &mut self.entries);
+    }
+
+    let mut seen = HashSet::new();
+
+    for (package, package_path) in &self.entries {
+      if let Some(bin_entries) = &package.bin {
+        match bin_entries {
+          deno_npm::registry::NpmPackageVersionBinEntry::String(script) => {
+            let name = default_bin_name(package);
+            if !seen.insert(name) {
+              // we already set up a bin entry with this name
+              continue;
+            }
+            set_up_bin_entry(
+              package,
+              name,
+              script,
+              package_path,
+              bin_node_modules_dir_path,
+            )?;
+          }
+          deno_npm::registry::NpmPackageVersionBinEntry::Map(entries) => {
+            for (name, script) in entries {
+              if !seen.insert(name) {
+                // we already set up a bin entry with this name
+                continue;
+              }
+              set_up_bin_entry(
+                package,
+                name,
+                script,
+                package_path,
+                bin_node_modules_dir_path,
+              )?;
+            }
+          }
+        }
+      }
+    }
+
+    Ok(())
+  }
+}
+
+// walk the dependency tree to find out the depth of each package
+// that has a bin entry, then sort them by depth
+fn sort_by_depth(
+  snapshot: &NpmResolutionSnapshot,
+  bin_entries: &mut [(NpmResolutionPackage, PathBuf)],
+) {
+  enum Entry<'a> {
+    Pkg(&'a NpmPackageId),
+    IncreaseDepth,
+  }
+  let mut want = bin_entries
+    .iter()
+    .map(|(p, _)| p.id.clone())
+    .collect::<HashSet<_>>();
+  let mut seen = HashSet::new();
+  let mut depths: HashMap<NpmPackageId, u64> =
+    HashMap::with_capacity(want.len());
+
+  let mut queue = VecDeque::new();
+  queue.extend(snapshot.top_level_packages().map(Entry::Pkg));
+  queue.push_back(Entry::IncreaseDepth);
+
+  let mut current_depth = 0u64;
+
+  while let Some(entry) = queue.pop_front() {
+    if want.is_empty() {
+      break;
+    }
+    let id = match entry {
+      Entry::Pkg(id) => id,
+      Entry::IncreaseDepth => {
+        current_depth += 1;
+        if queue.is_empty() {
+          break;
+        }
+        queue.push_back(Entry::IncreaseDepth);
+        continue;
+      }
+    };
+    if let Some(package) = snapshot.package_from_id(id) {
+      if want.remove(&package.id) {
+        depths.insert(package.id.clone(), current_depth);
+      }
+      seen.insert(package.id.clone());
```

Comment:
I think we should change this (and above) to insert to `seen` when adding to the queue. Otherwise the same `id` could be added multiple times and we'd be doing more work.

```rs
for dep_id in package.dependencies.values() {
  if seen.insert(dep_id) {
    queue.push(Entry::Pkg(dep_id));
  }
}
```

(Also, we're unconditionally adding stuff to the queue on the next line even if the current dep might have been seen before)